util/collections/src/garbage/map.rs

```diff
@@ -0,0 +1,131 @@
+use crate::garbage::Duration;
+use std::collections::{BTreeMap, HashMap};
+use std::fmt::Debug;
+use std::hash::Hash;
+
+pub struct GcMap<K, V>
+where
+	K: Eq + Hash + Debug,
+	V: Eq + Hash + Debug,
+{
+	/// The number of some unit time a value is valid for.
+	value_ttl: Duration,
+	/// The duration of a garbage collection slot in some unit time.
+	/// This is used to bin values into slots for O(value_ttl/gc_slot_duration * log value_ttl/gc_slot_duration) garbage collection.
+	gc_slot_duration: Duration,
+	/// The value lifetimes, indexed by slot.
+	value_lifetimes: BTreeMap<u64, HashMap<K, V>>,
+}
+
+impl<K, V> GcMap<K, V>
+where
+	K: Eq + Hash + Debug,
+	V: Eq + Hash + Debug,
+{
+	/// Creates a new GcMap with a specified garbage collection slot duration.
+	pub fn new(value_ttl: Duration, gc_slot_duration: Duration) -> Self {
+		GcMap { value_ttl, gc_slot_duration, value_lifetimes: BTreeMap::new() }
+	}
+
+	/// Gets a value for a key
+	pub fn get_value(&self, key: &K) -> Option<&V> {
```

Comment:
This can be simplified as
```rust
for lifetimes in self.value_lifetimes.values().rev() {
        if let Some(value) = lifetimes.get(key) {
            // Return immediately if a valid value is found
            return Some(value);
        }
    }
    None
````